File: LeetCode-Java/Longest Valid Parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        if(s == null) return 0;
        int m = s.length();
        if(m <= 1) return 0;
        int[] record = new int[m];
        record[0] = -1;
        for(int i = 1;i < m;i ++)
        {
            if(record[i - 1] == -1)
            {
                int tempResult = -1;
                if( s.charAt(i) == ')' && s.charAt(i - 1) == '(')
                {
                    tempResult = i - 1;
                    if(i - 2 >= 0 && record[i - 2] >= 0)
                        tempResult = record[i - 2];
                }
                record[i] = tempResult;
            }
            else
            {
                int index = record[i - 1] - 1;
                if(index < 0 || !(s.charAt(index) == '(' && s.charAt(i) == ')'))
                    record[i] = -1;
                else
                {
                    record[i] = index;
                    if(index - 1 >= 0 && record[index - 1] >= 0)
                        record[i] = record[index  - 1];
                }
            }
        }
        int maxLength = 0;
        for(int i = 0;i < m;i ++)
        {
            if(record[i] >= 0)
                maxLength = Math.max(maxLength, i - record[i] + 1);
        }
        return maxLength;
    }
}

File: Leetcode-cpp/longest_valid_parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> record;
        int result = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (s[i] == '(') {
                record.push(i);
            } else if (!record.empty() && s[record.top()] == '(') {
                record.pop();
            } else {
                record.push(i);
            }
        }
        int current = s.length() - 1;
        while (!record.empty()) {
            result = max(result, current - record.top());
            current = record.top() - 1;
            record.pop();
        }
        result = max(result, current + 1);
        return result;
    }
};